fix(compiler-cli): type-check inputs that include undefined when there's coercion members

JoostK opened the pull request:
For attribute bindings that target a directive's input, the template
type checker is able to verify that the type of the input expression is
compatible with the directive's declaration for said input. This
checking adheres to the `strictNullChecks` flag as configured in the
TypeScript compilation, such that errors are reported for expressions
that include `undefined` or `null` in their type if the input's
declaration does not include those types.

There was a bug with this level of type-checking for directives that
also declare coercion members, where binding an expression that includes
the `undefined` type to a directive's input that does not include the
`undefined` type would not be reported as error.

This commit fixes the bug by changing the type-constructor in type-check
code to use an intersection type of regular inputs and coerced inputs,
instead of a union type. The union type would inadvertently allow
`undefined` types to be assigned into the regular inputs, as that would
still satisfy the characteristics of a union type.

JoostK committed:
fix(compiler-cli): type-check inputs that include undefined when there's coercion members

For attribute bindings that target a directive's input, the template
type checker is able to verify that the type of the input expression is
compatible with the directive's declaration for said input. This
checking adheres to the `strictNullChecks` flag as configured in the
TypeScript compilation, such that errors are reported for expressions
that include `undefined` or `null` in their type if the input's
declaration does not include those types.

There was a bug with this level of type-checking for directives that
also declare coercion members, where binding an expression that includes
the `undefined` type to a directive's input that does not include the
`undefined` type would not be reported as error.

This commit fixes the bug by changing the type-constructor in type-check
code to use an intersection type of regular inputs and coerced inputs,
instead of a union type. The union type would inadvertently allow
`undefined` types to be assigned into the regular inputs, as that would
still satisfy the characteristics of a union type.

As a result of this change, you may start to see build failures if
`strictTemplates` is enabled and `strictInputTypes` is not disabled.
These errors are legitimate and some action is required to achieve a
successful build:

1. Update the templates for which an error is reported and introduce the
   non-null assertion operator at the end of the expression. This
   removes the `undefined` type from the expression's type, making it
   appear as a valid assignment.
2. Disable `strictNullInputTypes` in the compiler options. This will
   implicitly add the non-null assertion operators similar to option 1,
   but all templates in the compilation are affected.
3. Update the directive's input declaration to include the `undefined`
   type, if the directive is not implemented in an external library.

atscott commented:
[presubmit](https://test.corp.google.com/ui#id=OCL:323823951:BASE:323824019:1596045746186:3f0c8d92)

JoostK committed:
test: update components repo to test against recent revision

The changes in angular/components#20136 are required to allow the
framework tests to succeed.

atscott reviewed (changes requested):
@JoostK - Can you update the commit message to include a small note about how to get around this, either by using a non-null assertion operator or by turning off `strictNullInputTypes` (?). In addition, maybe adding another test to demonstrate that `strictNullInputTypes: false` (again, not sure if this is the correct option) does not produce an error in this case.

devversion reviewed (approved):
LGTM for dev-infra.

@JoostK When you get a chance (for later reference): could you briefly explain why we go from v7 to v9 in the cache key?

JoostK commented:
D'oh! I meant to do that but then completely forgot about it.

I used the v8 key for testing against the required commit using my fork, to verify whether the components repo PR would actually build successfully. Upon switching back to the primary repo, the v8 fallback cache was poisoned with a checkout of my fork.

AndrewKushnir commented:
Hi @JoostK, thanks for fixing the issue! 👍 

Quick question: should this PR be also included into the patch branch, since this is a fix (it has a `master-only` PR target atm)?

Thank you.

atscott commented:
@AndrewKushnir We merged to master only out of an abundance of caution. It's technically a breaking change because it can produce new errors that weren't detected before. Those errors _should_ have already been appearing though, so this is a bug fix but also breaking.

atscott commented:
@JoostK Speaking of breaking, should we have added the `BREAKING CHANGE:` footnote?

AndrewKushnir commented:
Thanks for additional information @atscott, that makes sense.

JoostK commented:
@AndrewKushnir the change to update the components repo commit should also be in patch, so I opened #38378 to land that in patch.

@atscott I intentionally did _not_ include an explicit breaking changes note, as we considered this as a true bug fix that can land anytime, without constraining ourselves to targeting it for a major. I guess we don't really have a proper communication channel to give people a heads up about these kinds of changes; the callout in the changelog would still be useful, but currently that can only be done using a breaking change note (AFAIK).

AndrewKushnir commented:
Thanks @JoostK 👍
